### PR TITLE
SNI support for OpenSSL::SSL::Socket and HTTP::Client

### PR DESCRIPTION
**Pull request by jhass**
I'm not sure whether / how to write specs for this :/

### CODE REVIEW

File: src/openssl/ssl/socket.cr [hunk 1]
```diff
@@ -29,14 +51,21 @@ class OpenSSL::SSL::Socket
...
   end
 
   def write(slice : Slice(UInt8))
     check_open
 
     count = slice.size
-    LibSSL.ssl_write(@ssl, slice.pointer(count), count)
+    bytes = LibSSL.ssl_write(@ssl, slice.pointer(count), count)
+    unless bytes > 0
+      raise OpenSSL::SSL::Error.new(@ssl, bytes, "SSL_read")
```
Comment (ysbaddaden) [inline src/openssl/ssl/socket.cr]: Copy paste typo: `SSL_write`

Comment (jhass) [inline src/openssl/ssl/socket.cr]: Fixed.

File: src/openssl/openssl.cr [hunk 2]
```diff
@@ -13,6 +13,45 @@ module OpenSSL
...
+          msg = "#{"#{func}: " if func}: Unknown error"
+        else
+          err_msg = String.new(LibCrypto.err_error_string(code, nil))
+          msg = "#{"#{func}: " if func}#{err_msg}"
+        end
+      else
+        msg = "#{"#{func}: " if func}#{@err}"
+      end
+
+      super(msg)
```
Comment (ysbaddaden) [inline src/openssl/openssl.cr]: Thank you OpenSSL. This is awful :/

Comment (jhass) [inline src/openssl/openssl.cr]: I know, and I'm not 100% certain it even handles all cases yet, but I hope so.

**Comment by ysbaddaden** [comment]
This looks great! Checking error messages is indeed better. The chosen method to verify the hostname seems backward compatible with at least OpenSSL 1.0.1, which is perfect.

Maybe we could test against the badssl.com domains? Thought it's only meant for _manual_ testing.

**Comment by jhass** [comment]
More importantly we shouldn't make the testsuite dependent on external network calls, as quite a few (distribution) build environments disallow external network connections.

**Comment by jhass** [comment]
Refactored `OpenSSL::SSL::Error` a little.

**Comment by asterite** [comment]
With these changes I now get:

> SSL_connect: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure (OpenSSL::SSL::Error)

for this program:

```crystal
require "http"

client = HTTP::Client.new "getmondo.co.uk", ssl: true
resp = client.get("/")
puts resp.body
```

Is this expected?

**Comment by jhass** [comment]
It works for me, what's your openssl version?

**Comment by asterite** [comment]
```
$ openssl version
OpenSSL 0.9.8zh 14 Jan 2016
```

**Comment by asterite** [comment]
Hm, I also linked a newer one and still the same error:

```
$ openssl version
OpenSSL 1.0.2h  3 May 2016
```

**Comment by jhass** [comment]
I run `1.0.2h` too, SNI was backported to `0.9.8f` so it's not just lacking support. Different default ciphers perhaps? Any idea how I could reproduce locally? Or could you create a Wireshark/pcap capture of that connection?

**Comment by jhass** [comment]
Couldn't reproduce with

* Archlinux, OpenSSL 1.0.2h
* Ubuntu 12.04, OpenSSL 1.0.1
* Ubuntu 16.04, OpenSSL 1.0.2g
* CentOS 6, OpenSSL 1.0.1e
* Debian Wheezy, OpenSSL 1.0.1e

**Comment by asterite** [comment]
Feel free to merge. It must be something wrong with my setup.

We could maybe later test this by starting an http server and using an http client, though maybe as an integration test.

**Comment by ysbaddaden** [comment]
Thanks again for adding the OpenSSL error checks. This really helps me for #2671 !

I'm adding them to OpenSSL::SSL::Context too.

**Comment by asterite** [comment]
This code:

```crystal
require "http/client"

puts HTTP::Client.get("https://www.google.com").body
```

Now gives me:

```
SSL_shutdown: Resource temporarily unavailable (Errno)
[4349575138] *CallStack::unwind:Array(Pointer(Void)) +82
[4349575041] *CallStack#initialize<CallStack>:Array(Pointer(Void)) +17
[4349575000] *CallStack::new:CallStack +40
[4349592969] *Exception@Exception#initialize<Errno, String>:CallStack +41
[4349592899] *Errno#initialize<Errno, String>:CallStack +115
[4349592748] *Errno::new<String>:Errno +108
[4349660866] *OpenSSL::SSL::Error#initialize<OpenSSL::SSL::Error, Pointer(Void), Int32, String>:CallStack +418
[4349660426] *OpenSSL::SSL::Error::new<Pointer(Void), Int32, String>:OpenSSL::SSL::Error +138
[4349659158] *OpenSSL::SSL::Socket#close<OpenSSL::SSL::Socket>:Nil +182
[4349694554] *HTTP::Client#close<HTTP::Client>:Nil +122
[4349691540] *HTTP::Client::exec<String, String, Nil, Nil>:HTTP::Client::Response +612
[4349690911] *HTTP::Client::get<String>:HTTP::Client::Response +31
[4349502470] __crystal_main +1174
[4349545016] main +40
```

**Comment by asterite** [comment]
Seems to be just a problem with SSL_shutdown, so should be simple to fix. The docs say:

```
RETURN VALUES
       The following return values can occur:

       0   The shutdown is not yet finished. Call SSL_shutdown() for a second time, if a bidirectional shutdown shall be performed.  The output of SSL_get_error(3) may be misleading,
           as an erroneous SSL_ERROR_SYSCALL may be flagged even though no error occurred.

       1   The shutdown was successfully completed. The "close notify" alert was sent and the peer's "close notify" alert was received.

       <0  The shutdown was not successful because a fatal error occurred either at the protocol level or a connection failure occurred. It can also occur if action is need to
           continue the operation for non-blocking BIOs.  Call SSL_get_error(3) with the return value ret to find out the reason.
```

> It can also occur if action is need to continue the operation for non-blocking BIOs

Maybe `< 0` doesn't always mean an error.

**Comment by asterite** [comment]
Also found: http://stackoverflow.com/a/31613293

**Comment by asterite** [comment]
And also: http://stackoverflow.com/a/28056464

I don't understand why C APIs are sometimes so complex :(

**Comment by jhass** [comment]
@asterite can you check if d5fe700ee0dfa6856364f31650fb2331dcbfdaf2 fixed it for you?

**Comment by asterite** [comment]
@jhass Fixed! :-)

I left a comment about not using exceptions for control flow

**Comment by ysbaddaden** [comment]
Note: C may be complex sometimes, but in this case there are 2 things:

1. detect whether a bidirectionnal shutdown was initiated (this is a SSL/TLS complexity);
2. avoid blocking the whole process in case of a bidirectionnal shutdown. Since we have the `SSL_ERROR_WANT_READ` and `SSL_ERROR_WANT_WRITE` error codes, we could ask libevent to resume the fiber when the underlying socket is ready to read or write, then try the shutdown again

I say kudos to OpenSSL (for once) for providing an event-loop friendly API.

**Comment by jhass** [comment]
Yes, however in my tests those weren't actually returned, but `SYSCALL` was and the underlying `Errno::EAGAIN` was set.

**Comment by ysbaddaden** [comment]
Damn, I shouldn't have expected that much. Maybe we should just `Fiber.yield` then :D

**Comment by asterite** [comment]
IO::FileDescriptor has `wait_readable` and `wait_writeable` methods to do that, just that they are private. Maybe this should be public but `:nodoc:` or similar.

**Pull request merged**